text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with build-time and runtime Dockerfiles that act as pre-build hooks.
+These hooks would be equivalent to base image / builder customizations, but would execute immediately before a build.
+
+Most importantly: the changes below do not impact buildpacks or the buildpack API.
+For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder with the hooks pre-applied, without modifying the buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "hook" files in `/cnb/hooks.d/`. Hooks may be associated with zero or more buildpack IDs.
+The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
+```toml
+[[hooks]]
+name = "app-hook"
+path = "./myhook"
```

Comment:
Why do we need `path` if there's such a strong specification of the file path/name/extension (below)?

Reply:
Ok, now I believe this is the path that's copied into the builder, at the path/filename defined below.

Reply:
Alternative: Instead of exposing the "hook" concept to builder-image creators, we use the formats we support explicitly as tables under the base image. For example (assuming `[stack]` goes away):

```
[base-image.dockerfile]
path = "<string (path to file)>"
run = "<boolean>"
build = "<boolean>"
out = "<boolean>"
buildpacks = ["<string (buildpack ID)>"]
```

I'm open to other ideas that avoid introducing "hooks" to builder-image owners.

Reply:
You lose flexibility. A hook can dynamically generate a Dockerfile. It could do that based on any information it has at the time it runs, such as the contents of the user's application or what buildpacks have passed detection. If you remove the hook then the builder can only add Dockerfiles that are set at the time the builder is created.

As an example, without the hook, you couldn't implement the [use case from the RFC](https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R84-R93), where the hook takes the build and run Dockerfiles it generates from the app directory.

Reply:
My example would dynamically generate a Dockerfile if `out = true`

Reply:
Basically, I'm not trying to change the mechanism described below (hook.d is fine in the internals). It's how it manifests in the dev ex that concerns.

Reply:
Actually, I question the value of `out = false` or a hook w/o `.out`. If we were to eliminate the static Dockerfile capability (i.e. if you want that go run `docker build` before `pack builder create`), then it might open up some more (and potentially simpler) UX options

Reply:
Sorry, I'm not sure I follow. My present understanding is that like 90% of what's under [this section](https://github.com/buildpacks/rfcs/blob/dockerfiles/text/0000-dockerfiles.md#dynamically-applied-dockerfiles) is describing the internals. Most people wouldn't have to care about that, just the folks implementing this for tools like `pack`.

The actual user-facing part, that builder-builders would interact with is a.) the `[[hooks]]` table defined in the spec to indicate what hooks should be added to the builder image and b.) the hook itself.

Am I missing something?

>(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)

Or are you just trying to flush out this bit (ie. the format of `[[hooks]]`)? ^^

Reply:
`[[hook]]` is a concept that a builder image owner would need to learn and understand it. Most likely there would be overlap with buildpack authors needing to understand it.

Reply:
So is your concern with the name `hook`? Do you think it needs a different descriptive word?

The example you gave seems misleading to me, `[base-image.dockerfile]` implies that you'd have a static Dockerfile, not something that's executable. I think that was part of my confusion/question above. 

To me something like `[base-image.modifier]` would make more sense. It avoids the name hook. You could then have flags like `executable=true|false` and `format=dockerfile` (or whatever other formats might be supported in the future).

Reply:
Yea, I admit my alternative is misleading (implies a static Dockerfile). But generally, I'm wondering if we can up-level the `format` such that the only construct you need to understand is the format of that artifact (i.e. the hook becomes an impl detail of how that thing is run).